thrift/client.go

```diff
@@ -155,6 +157,37 @@ func (c *client) Call(ctx Context, thriftService, methodName string, req, resp t
 		return false, err
 	}
 
-	ctx.SetResponseHeaders(respHeaders)
+	ctx.SetResponseHeaders(removeRPCServiceHeader(respHeaders))
 	return isOK, nil
 }
+
+// removeRPCServiceHeader removes the $rpc$-service from headers map when present
+// tchannel inbound does not add a $rpc$-header. tchannel outbound sets the response headers
+// that are received from the outbound call on to the context and they get sent back to the
+// caller of this inbound. This becomes problematic when a tchannel inbound server is an
+// intermediate part of a call graph where the initial callee verifies $rpc$-service header
+// Eg. A -> B -> C
+//           |-> D
+// In this call graph, let us say B is a tchannel inbound server and makes two tchannel outbound
+// calls to C and D. If C and D both reply with response headers, the headers get set in B's
+// context, one overwriting the other, and they get sent back to A. If either one of C or D or both
+// send a $rpc$-header, A will start receiving one of C's or D's $rpc$-service response header and
+// can perform validation on the header value and can inadvertently believe that some other service
+// other B handled the request that it sent to B and can mark the rpc as failure. This is the behavior
+// that is observed with yarpc/yarpc-go framework when one of the intermediaries in the call graph
+// is using a tchannel inbound server using a tchannel outbound.
+//
+// We have observed this issue while migrating to yarpc from tchannel inbound. For example, let us
+// assume that both C & D were tchannel inbounds earlier and one of them wants to migrate to yarpc
+// and at which point the $rpc$-service header flows upstream and assuming A is already using yarpc outbound
+// the call will fail. Although it is a backward incompatible change, by removing the $rpc$-service header
+// prevents A from seeing the $rpc$-header and marking the rpc as a failure. The alternative is to make
+// B use Child() context while making downstream calls and the net effect is the same and while migrating to yarpc,
+// it is unreasonable to expect the callee service owners to perform changes into the upstream callers
+// to create Child() context especially when the call graph is complex and multiple levels deep. This is
+// a reasonbable backward incompatible change to make as we do not expect any caller to actually depend on
+// or find the $rpc$-service header from a descendant service to be actually useful.
+func removeRPCServiceHeader(headers map[string]string) map[string]string {
```

Comment:
separately, can/should we call this `removeSingleHopHeaders`, or is it too early? I imagine there's more of these - one example I can think of already is `load`. 

Also, perhaps unrelated to this change, I believe we hit a very similar issuer with service auth, where a tchannel service was incorrectly copying inbound single-hop tokens to their outbounds.  Similar, but looks like a different code path?

Reply:
+1. tchannel sits a layer below yarpc, so coupling in yarpc semantics (e.g. `$rpc$-` headers) breaks the abstraction. prefer to implement this as a configurable list of headers that should be stripped by the client, rather than hard-coding specific headers. since tchannel upgrades are a real pain in the monorepo, this can also help alleviate unnecessary overhead if other headers also need to be stripped.